libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,509 @@
+erased more often the others.
+
+The other problem is the reserved region where we would move data to. This
+would cost space as an entire region would be dedicated to garbage collecting.
+The garbage collecting region would also have more erase and writes performed
+on it breaking the wear leveling requirement.
+
+### Somewhat high storage overhead
+
+The storage overhead is somewhat high for TickFS. This is mostly due to the
+two 64-bit hashes that are stored with every object.
+
+The 64-bit value is determined by the output of the standard Rust
+`core::hash::Hasher` trait.
+
+For the hash of the key a 64-bit value can be justified by the lack of
+collision avoidance in the implementation. If two keys have the same
+hash the second key will be dropped. In this case a 64-bit hash should
+hopefully make that occurrence very unlikely. Also by following the standard
+Rust `core::hash::Hasher` trait the user is free to implement any standard
+Hasher of their choosing. Some systems will even be able to offload the
+hash to hardware.
+
+As for the check-sum. It was seriously considered to change this to a
+32-bit or even 16-bit CRC. This would reduce the overhead for each object
+by 4 or 6 bytes. Although a CRC is relatively easy to implement, it can still
+be time consuming. For this reason lots of embedded hardware will include a
+CRC offload engine to offload the calculation. By implementing a CRC in this
+library I would remove the option of allowing a offload engine. As there
+doesn't seem to be a generic CRC Rust trait to take advantage of the decision
+was made to reuse the `core::hash::Hasher` trait for the checksum. This costs
+space, but hopefully reduces implementation burden and allows for more hardware
+offloading.
+
+### Memory usage
+
+At a minimum TickFS requires a `PageSize` amount of free memory to store the
+buffer obtained when reading flash. This could be reduced by instead performing
+multiple reads of a smaller fixed size.
+
+### Synchronous and Asynchronous Usage
+
+TickFS supports both an synchronous and asynchronous usage. For developers
+used to Tock's callback method the the async usage might seem strange and
+sub-optimal. There are a few reasons that it is done this way.
+
+One of the goals of TickFS is to allow other non Tock users to use it. The
+current method should allow the library to be used to crates.io by other
+developers.
+
+Another reason for the synchronous support is that generally when there is
+only a single flash bank all flash writes and erases must be done synchronously.
+The synchronous API should allow these type of operations to occur.
+
+On top of that it is much simpler to unit test a synchronous API. This means the
+TickFS implementation can take advantage of a large range of unit tests to
+ensure correctness in the design. This is more difficult and complex to do in
+a async only operation.
+
+The async functions (see the documentation) can be used to implement a fully
```

Comment:
If you want TickFS to be completely general for the type of NV storage it's on top of, having async and sync APIs makes sense. For example, FRAM is close enough to RAM speeds (and is usually integrated with RAM e.g. in terms of addressing) that it's not useful to have an async API.

Can this document describe the API and how this works?

Why can't the callback that tells a client that an async operation has completed be part of the API? It could be a client doesn't bother registering these callbacks, or a synchronous implementation doesn't ever issue them. I think it's very important that this callback originate from the same module, at least -- you shouldn't have to register a callback with some board-and-TickFS-implementation-specific callback to be able to use TickFS.

Do you retry an operation because it wasn't complete? Or because the implementation was busy?

Reply:
This is the spec documetnation, it describes the underlying spec and not the usage interface. The information for how to use the API is in the rust docs.

Do you mean add a Tock callback function to the API that is then set as a client in the kernel?

We retry because it wasn't complete.